binaries/coordinator/src/lib.rs

```diff
@@ -39,15 +36,13 @@ mod run;
 mod tcp_utils;
 
 pub async fn start(
-    port: Option<u16>,
+    bind: SocketAddr,
     external_events: impl Stream<Item = Event> + Unpin,
-) -> Result<(u16, impl Future<Output = eyre::Result<()>>), eyre::ErrReport> {
-    let port = port.unwrap_or(DORA_COORDINATOR_PORT_DEFAULT);
-    let listener = listener::create_listener(port).await?;
-    let port = listener
+) -> Result<(SocketAddr, impl Future<Output = eyre::Result<()>>), eyre::ErrReport> {
+    let listener = listener::create_listener(bind).await?;
+    let bound_addr = listener
         .local_addr()
```

Comment:
I think we have to be careful here as we're getting the local address of the coordinator and not its current set address.

Reply:
Yes, we only returned the port here deliberately. Daemons need the public IP of the coordinator to create a connection, which we don't know.

Reply:
Isn't that what the current `bind` argument is about?

Reply:
As I understand it, the `bind` argument is always a local interface address. You can set it to 0.0.0.0 as a wildcard or pass a `bind` argument to choose one specific interface only. In order to connect daemons to the coordinator, they need to know the public IP assigned to the coordinator machine, as they can't connect to 0.0.0.0. This public IP might not even be known to the coordinator machine as it could be assigned and managed by some other network infrastructure (e.g. routers, NATs). You can tell daemons the public coordinator addr through the `--coordinator-addr` argument.

Reply:
Does the same network topology uncertainty apply to the daemons - i.e. the IP address that the coordinator receives when a daemon connects to it isn't necessarily the same IP address another daemon can use to initiate a connection?

Reply:
Yes, they could be different. I'm not sure how common this this in practice though. What's more likely is that the port number is remapped by some router, e.g. through a NAT protocol.

I'm not sure if it's a good idea to try to fix all of these things ourselves. It's probably better to use some middleware again instead of our custom TCP stream. For example, we could try to use `zenoh` again or the try the `tower` project.